File: scripts/create_hdf5.R

library(rhdf5)
library(Matrix)
library(org.Hs.eg.db)
library(org.Mm.eg.db)
library(AnnotationDbi)
library(data.table)
write_hdf5 <- function(path, counts_file, organism, assembly) {
  #Read in counts file
  expr <- as.data.frame(fread(counts_file, sep = "\t", header = TRUE))
  rownames(expr) <- expr$GENE
  expr <- expr[, 2:ncol(expr)]

  #Create Files for hdf5 write
  barcodes <- colnames(expr)
  gene_name <- rownames(expr)
  matrix <- Matrix(expr, sparse = TRUE)
  feature_type <- "Gene Expression"
  #Create Ensembl Symbol group
  if (organism == "mouse") {
    gene_symbol <- mapIds(org.Mm.eg.db,
                          keys = gene_name,
                          column = "ENSEMBL",
                          keytype = "SYMBOL",
                          multiVals = "first")
    } else {
    gene_symbol <- mapIds(org.Hs.eg.db,
                          keys = gene_name,
                          column = "ENSEMBL",
                          keytype = "SYMBOL",
                          multiVals = "first")
                          }

  #Create HDF5 File
  path <- path.expand(path)
  h5createFile(path)
  group <- "matrix"
  #Write Matrix Information to HDF5 File
  h5createGroup(path, group)
  h5write(barcodes,
          file = path,
          name = paste0(group, "/barcodes"))
  h5write(matrix@x,
          file = path,
          name = paste0(group, "/data"))
  h5write(matrix@i,
          file = path,
          name = paste0(group, "/indices"))
  h5write(matrix@Dim,
          file = path,
          name = paste0(group, "/shape"))
  h5write(matrix@p,
          file = path,
          name = paste0(group, "/indptr"))

  #Write Feature info in hdf5 file
  h5createGroup(path, file.path(group, "features"))
  h5write(gene_symbol,
          file = path,
          name = paste0(group, "/features/id"))
  h5write(gene_name,
          file = path,
          name = paste0(group, "/features/name"))
  h5write(rep(feature_type, length.out = length(gene_name)),
          file = path,
          name = paste0(group, "/features/feature_type"))
  h5write("genome",
          file = path,
          name = paste0(group, "/features/_all_tag_keys"))
  h5write(rep(assembly, length.out = length(gene_name)),
          file = path,
          name = paste0(group, "/features/genome"))
}

#CREATE OPTION PARSER FOR HDF5 FUNCTION NEES
library(optparse)
option_list <- list(
  make_option(c("-i", "--input"),
    type = "character",
    default = NULL,
    help = "count matrix .tsv format (gene by cell matrix)",
    metavar = "character"),

  make_option(c("-o", "--out"),
    type = "character",
    default = "out.hdf5",
    help = "output file to store hdf5 structure",
    metavar = "character"),

  make_option(c("-s", "--species"),
    type = "character",
    default = NULL,
    help = "genome species (mouse or human)",
    metavar = "character"),

  make_option(c("-v", "--assembly_version"),
    type = "character",
    default = NULL,
    help = "genome assembly version",
    metavar = "character")
)
#GENERATE PARSER FROM COMMDANDLINE INPUTS
opt_parser <- OptionParser(option_list = option_list)
opt <- parse_args(opt_parser)

#CALL HDF5 FUNCTION
write_hdf5(path = opt$out,
  counts_file = opt$input,
  organism = opt$species,
  assembly = opt$assembly_version)

File: xenomake/postprocess/create_hdf5.py
"""
Author: Ivy Strope
Created: 8/18/23
Contact: u247529@bcm.edu
Last Edited: 8/18/23 
"""
"""
Author: Ivy Strope
Created: 8/18/23
Contact: u247529@bcm.edu
Last Edited: 8/18/23
"""
#!/usr/bin/env python
import h5py
import argparse
import pandas as pd
import numpy as np
from scipy.sparse import csr_array

#SET UP HIERARCHICAL DATA STRUCTURE THAT MIMIC 10X OUTPUT
def create_hdf5(path,counts,organism,assembly):
    #Process Counts File
    data = pd.read_csv(counts,sep='\t',index_col='GENE')
    mtx = csr_array(data,dtype=int)  #convert to sparse format
    gene_names = list(data.index) #extract gene names
    ensembl = map_to_ensembl(genes = gene_names,organism = organism) #convert to ensembl id
    barcodes = list[data.columns.values] #extract spot barcode info
    feature_type = ['Gene_Expression'] * len(gene_names)

    #Create HDF5 File
    f = h5py.File(path,'w')
    #Create Hierarchical organization
    matrix = f.create_group('matrix')
    features = f.create_group('matrix/features')

    #Fill out matrix group
    matrix.create_dataset(name='barcodes',data=np.array(barcodes,dtype='S'))
    matrix.create_dataset(name='matrix',data=mtx.data,dtype='i')
    matrix.create_dataset(name='indices',data=mtx.indices,dtype='i')
    matrix.create_dataset(name='indptr',data=mtx.indptr,dtype='i')
    matrix.create_dataset(name='shape',data=mtx.shape,dtype='i')

    #Fill out features group
    features.create_dataset(name='_all_tag_keys',data = np.array('genome',dtype='S'))
    features.create_dataset(name = 'feature_type',
                                    data = np.array(feature_type,dtype='S'))
    features.create_dataset(name = 'genome',data = np.array(assembly,dtype='S'))
    features.create_dataset(name = 'id', data = np.array(ensembl,dtype='S'))
    features.create_dataset(name = 'name', data = np.array(gene_names,dtype='S'))

    f.close()


#MAP BETWEEN GENE SYMBOL AND ENSEMBL ID (Required for 10x)
def map_to_ensembl(genes,organism):

    if (organism == 'human'):
        gtf = pd.read_csv('species/human/ids.txt',sep='\t')
    else:
        gtf = pd.read_csv('species/mouse/ids.txt',sep='\t')
    gtf = gtf.reindex(genes)
    id = list(gtf['Geneid'])
    return id



parser = argparse.ArgumentParser(
        allow_abbrev=False,
        add_help=False)

parser.add_argument(
        '--counts',
        help='digital gene expression output file',
        type=str,
        required=True)

parser.add_argument(
        '--species',
        help='specify if mouse or human species',
        type=str,
        required=True)

parser.add_argument(
        '--assembly',
        help='specify assembly version',
        type=str,
        required=True)
parser.add_argument(
        '--output',
        help='output file path',
        type=str,
        required=True)


#RUN SCRIPTS
args = parser.parse_args()
create_hdf5(path = args.output,
        counts = args.counts,
        organism = args.species,
        assembly = args.assembly)
